src/psm/src/shape.cpp

```diff
@@ -0,0 +1,296 @@
+
+#include "shape.h"
+
+#include <boost/geometry.hpp>
+#include <boost/polygon/polygon.hpp>
+
+#include "node.h"
+
+namespace psm {
+
+Shape::Shape(const odb::Rect& shape, odb::dbTechLayer* layer)
+    : shape_(shape), layer_(layer)
+{
+}
+
+odb::dbTechLayer* Shape::getLayer() const
+{
+  return layer_;
+}
+
+std::vector<std::unique_ptr<Connection>> Shape::connectNodes(
+    const IRNetwork::NodeTree& layer_nodes)
+{
+  std::vector<std::unique_ptr<Connection>> shape_connections;
+
+  Node::NodeSet used;
+
+  const auto sorted_nodes = getNodes(layer_nodes);
+  const auto tree = getNodeTree(sorted_nodes);
+
+  for (auto* node : sorted_nodes) {
+    const auto& pt = node->getPoint();
+    const IRNetwork::Point point(pt.x(), pt.y());
+
+    std::vector<IRNetwork::NodeValue> ordered_neighbors;
+
+    used.insert(node);
+
+    tree.query(boost::geometry::index::satisfies([&](const auto value) {
+                 return used.find(value.second) == used.end();
+               }) && boost::geometry::index::nearest(point, 1),
+               std::back_inserter(ordered_neighbors));
+
+    for (const auto& [other_pt, other] : ordered_neighbors) {
+      const int len_x
+          = std::abs(other->getPoint().getX() - node->getPoint().getX());
+      const int len_y
+          = std::abs(other->getPoint().getY() - node->getPoint().getY());
+
+      int len;
+      int width;
+      if (len_x > len_y) {
+        len = len_x;
+        width = shape_.dy();
+      } else {
+        len = len_y;
+        width = shape_.dx();
+      }
+
+      shape_connections.push_back(
+          std::make_unique<LayerConnection>(node, other, len, width));
+    }
+  }
+
+  return shape_connections;
+}
+
+std::string Shape::describe(double dbu) const
+{
+  return fmt::format("{}: ({:.4f}, {:.4f}) -- ({:.4f}, {:.4f})",
+                     id_,
+                     shape_.xMin() / dbu,
+                     shape_.yMin() / dbu,
+                     shape_.xMax() / dbu,
+                     shape_.yMax() / dbu);
+}
+
+std::vector<std::unique_ptr<Node>> Shape::createFillerNodes(
+    int max_distance,
+    const IRNetwork::NodeTree& layer_nodes)
+{
+  std::vector<std::unique_ptr<Node>> new_nodes;
+
+  int delta_x = 0;
+  int delta_y = 0;
+
+  const int radius = max_distance / 2;
+
+  odb::Point start;
+  if (shape_.dx() > shape_.dy()) {
+    delta_x = max_distance;
+    start = odb::Point(shape_.xMin() + radius, shape_.yCenter());
+  } else {
+    delta_y = max_distance;
+    start = odb::Point(shape_.xCenter(), shape_.yMin() + radius);
+  }
+
+  const IRNetwork::NodeTree tree = getNodeTree(getNodes(layer_nodes));
+
+  while (shape_.overlaps(start)) {
+    const IRNetwork::Box box(
+        IRNetwork::Point(start.x() - radius, start.y() - radius),
+        IRNetwork::Point(start.x() + radius, start.y() + radius));
+
+    if (tree.qbegin(boost::geometry::index::intersects(box)) == tree.qend()) {
+      new_nodes.push_back(std::make_unique<Node>(start, layer_));
```

Comment:
This has the possibility to have nodes more than max_distance apart.  For example image a wire from [0,100] with a max_distance of 10.  If you have existing nodes at 6 (in [5,15]) and 24 (in [15,25]) you won't add a node but they are 18 apart.

Reply:
removed this logic